language/testing-infra/transactional-test-runner/src/tasks.rs

```diff
@@ -0,0 +1,317 @@
+            let data = NamedTempFile::new()?;
+            data.reopen()?
+                .write_all(file_text_vec.join("\n").as_bytes())?;
+            Some(data)
+        };
+
+        tasks.push(TaskInput {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        })
+    }
+    Ok(tasks)
+}
+
+impl<T> TaskInput<T> {
+    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> TaskInput<U> {
+        let Self {
+            command,
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        } = self;
+        TaskInput {
+            command: f(command),
+            name,
+            number,
+            start_line,
+            command_lines_stop,
+            stop_line,
+            data,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SyntaxChoice {
+    Source,
+    IR,
+}
+
+#[derive(Debug, StructOpt)]
+pub struct PublishCommand {
+    #[structopt(long = "gas-budget")]
+    pub gas_budget: Option<u64>,
+    #[structopt(long = "syntax")]
+    pub syntax: Option<SyntaxChoice>,
+    #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+    pub address: Option<AccountAddress>,
+}
+
+#[derive(Debug, StructOpt)]
+pub struct RunCommand {
+    #[structopt(long = "signers", parse(try_from_str = parse_account_address))]
+    pub signers: Vec<AccountAddress>,
+    #[structopt(long = "args", parse(try_from_str = parser::parse_transaction_argument))]
+    pub args: Vec<TransactionArgument>,
+    #[structopt(long = "type-args", parse(try_from_str = parser::parse_type_tag))]
+    pub type_args: Vec<TypeTag>,
+    #[structopt(long = "gas-budget")]
+    pub gas_budget: Option<u64>,
+    #[structopt(long = "syntax")]
+    pub syntax: Option<SyntaxChoice>,
+    #[structopt(name = "NAME", parse(try_from_str = parse_qualified_module_access))]
+    pub name: Option<(ModuleId, Identifier)>,
+}
+
+#[derive(Debug, StructOpt)]
+pub struct ViewCommand {
+    #[structopt(long = "address", parse(try_from_str = parse_account_address))]
+    pub address: AccountAddress,
+    #[structopt(long = "resource", parse(try_from_str = parse_qualified_module_access_with_type_args))]
+    pub resource: (ModuleId, Identifier, Vec<TypeTag>),
+}
+
+#[macro_export]
+macro_rules! define_commands {
+    ($command:ident
+        $(, publish:$expub:ident)?
+        $(, run:$exrun:ident)?
+        $(, subcommands:$subcommand:ident)?
+    ) => {
+        #[allow(unused_imports)]
+        use structopt::*;
+        #[derive(Debug, StructOpt)]
+        #[structopt(
+            name = "task",
+            about = "Transactional task commands",
+            rename_all = "kebab-case"
+        )]
+        pub enum $command {
+            #[structopt(name = "publish")]
+            Publish($crate::tasks::PublishCommand $(, $expub:ident)?),
```

Comment:
This doesn't seem to work. First, the `:ident` part should be removed. Second, after removing `:ident`, I got the following error when trying to add an extra argument:
```
error: non single-typed tuple enums are not supported
   --> language/testing-infra/transactional-test-runner/src/tasks.rs:245:13
    |
245 | /             #[structopt(name = "publish")]
246 | |             Publish($crate::tasks::PublishCommand $(, $expub)?),
    | |_______________________________________________________________^
    | 
   ::: language/testing-infra/transactional-test-runner/src/vm_test_harness.rs:27:1
    |
27  |   define_commands!(TaskCommand, publish: u64);
    |   -------------------------------------------- in this macro invocation
```

Reply:
Good call, fixed